File: packages/ts/src/domain/conversation.ts
import type { ToolCallId } from "./identifiers";
import type { ToolCall } from "./tool";

export const MessageRole = {
  User: "user",
  Assistant: "assistant",
  Tool: "tool",
} as const;
export type MessageRole = (typeof MessageRole)[keyof typeof MessageRole];

export interface UserMessage {
  readonly role: typeof MessageRole.User;
  readonly content: string;
}

export interface AssistantMessage {
  readonly role: typeof MessageRole.Assistant;
  readonly content: string;
  readonly toolCalls: readonly ToolCall[];
}

export interface ToolMessage {
  readonly role: typeof MessageRole.Tool;
  readonly toolCallId: ToolCallId;
  readonly content: string;
}

export type Message = UserMessage | AssistantMessage | ToolMessage;

File: packages/py/src/airlock/domain/conversation.py
"""Conversation messages exchanged with the model."""

from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from airlock.domain.identifiers import ToolCallId
from airlock.domain.tool import ToolCall


class MessageRole(StrEnum):
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


class UserMessage(BaseModel):
    model_config = ConfigDict(frozen=True)

    role: Literal[MessageRole.USER] = MessageRole.USER
    content: str


class AssistantMessage(BaseModel):
    model_config = ConfigDict(frozen=True)

    role: Literal[MessageRole.ASSISTANT] = MessageRole.ASSISTANT
    content: str
    tool_calls: tuple[ToolCall, ...] = ()


class ToolMessage(BaseModel):
    model_config = ConfigDict(frozen=True)

    role: Literal[MessageRole.TOOL] = MessageRole.TOOL
    tool_call_id: ToolCallId
    content: str


Message = Annotated[
    UserMessage | AssistantMessage | ToolMessage,
    Field(discriminator="role"),
]
